portable/GCC/ARM_CRx_MPU/mpu_wrappers_v2_asm.S

```diff
@@ -0,0 +1,1334 @@
+ * FreeRTOS Kernel <DEVELOPMENT BRANCH>
+ * Copyright (C) 2021 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+/*---------------------------------------------------------------------------*/
+
+    .arm
+    .syntax unified
+    .section freertos_system_calls
+#define FREERTOS_ASSEMBLY
+    #include "FreeRTOSConfig.h"
+    #include "portmacro_asm.h"
+    #include "mpu_syscall_numbers.h"
+#undef FREERTOS_ASSEMBLY
+
+/* ------------------ Start of Port Specific System Calls ------------------ */
+
+/**
+ * Function: void vPortYield
+ * Inputs:      VOID
+*/
+.align 4
+.global vPortYield
+.type vPortYield, %function
+vPortYield:
+    /* Make the SVC to swap tasks */
+    SVC     #portSVC_YIELD
+    /* After yielding to another task, resume executing the calling task */
+    BX  LR
+
+
+/*-------------------------------------------------------------------------------*/
+/* vPortSystemCallExit */
+.align 4
+.global vPortSystemCallExit
+.type vPortSystemCallExit, %function
+vPortSystemCallExit:
+    /* Make the SVC to exit a system call */
+    SVC     #portSVC_SYSTEM_CALL_EXIT
+    /* After performing the requested FreeRTOS API, return to the calling task */
+    BX      LR
+
+/*-------------------------------------------------------------------------------*/
+/**
+ * Function: BaseType_t xPortIsPrivileged
+ * Inputs:      VOID
+*/
+.align 4
+.weak xPortIsPrivileged
+.type xPortIsPrivileged, %function
+xPortIsPrivileged:
+    /* Load value of SPSR into R0 */
+    MRS     R0, CPSR
+    /* Get the relevant mode bits */
+    AND     R0, R0, #0x1F
+    /* Check if we're in user mode */
+    CMP     R0, #USER_MODE
+    /* If we are in user mode set R0 to 0 for the return */
+    MOVEQ   R0, #0x0
+    /* Otherwise set R0 to 1 for the return */
+    MOVNE   R0, #0x01
```

Comment:
Seems like it would be better to check if the R0 register is equal to the system mode (#0x1F) and return true if it is, this way all other modes are treated as user mode.

The update you would need:
```
CMP R0, #SYSTEM_MODE
MOVEQ R0, #0x1
MOVENE R0, #0x0
```

Reply:
Or is the goal here to treat all other modes as system modes?